### PR TITLE
[SDK-4319] Add support for Edge runtime

### PR DESCRIPTION
**Pull request by adamjmcgrath**
### 📋 Changes

Have added support for the Edge runtime (in the app directory) to the auth handlers and session helpers:

**Auth Handlers**
`handleAuth`
`handleLogin`
`handleCallback`
`handleLogout`
`handleProfile`

**Session Helpers**
`withApiAuthRequired`
`withPageAuthRequired`
`updateSession`
`getAccessToken`
(`getSession` is already supported)

We now use 2 OIDC clients:
- For Node: https://github.com/panva/node-openid-client
- For Edge: https://github.com/panva/oauth4webapi

Using Jest "projects" to run the handler tests in both runtimes
Have updated the example app to use edge as well as node
Also support added for middleware responses (see https://github.com/auth0/nextjs-auth0/issues/1150)

### 📎 References

Fixes https://github.com/auth0/nextjs-auth0/issues/912 https://github.com/auth0/nextjs-auth0/issues/1150 https://github.com/auth0/nextjs-auth0/issues/1257

### 🎯 Testing

There is an example app running on the edge runtime on Vercel https://app-dir-auth0.vercel.app/ ([source](https://github.com/adamjmcgrath/app-dir-auth0/blob/main/app/api/auth/%5Bauth0%5D/route.ts#L9))

### CODE REVIEW

File: example-app/app/nav.tsx [hunk 1]
```diff
@@ -41,28 +41,52 @@ export default function Nav() {
                 <a>Profile</a>
               </Link>
             </li>
+            <li>
+              <Link href="/edge-profile" legacyBehavior>
```
Comment (Widcket) [inline example-app/app/nav.tsx:45]: Is `legacyBehavior` still needed here (and in other links)?

File: src/auth0-session/client/edge-client.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+  private async getClient(): Promise<[oauth.AuthorizationServer, oauth.Client]> {
+    if (this.as) {
+      return [this.as, this.client as oauth.Client];
+    }
+
+    const issuer = new URL(this.config.issuerBaseURL);
+    try {
+      this.as = await oauth
+        .discoveryRequest(issuer, this.httpOptions())
+        .then((response) => oauth.processDiscoveryResponse(issuer, response));
```
Comment (Widcket) [inline src/auth0-session/client/edge-client.ts:73]: Is the discovery doc being cached?

File: src/auth0-session/client/edge-client.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+      Object.entries(extraParams).forEach(([key, value]: [string, string]) => {
+        if (value === null || value === undefined) {
+          return;
+        }
+        auth0LogoutUrl.searchParams.set(key, value);
+      });
+      return auth0LogoutUrl.toString();
+    }
+    if (!as.end_session_endpoint) {
+      throw new Error('no rp inititated logout');
```
Comment (Widcket) [inline src/auth0-session/client/edge-client.ts:188]: Should this error have a more specific type?

File: src/auth0-session/client/edge-client.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    const [as, client] = await this.getClient();
+
+    const { clientAssertionSigningKey, clientAssertionSigningAlg } = this.config;
+
+    let clientPrivateKey = clientAssertionSigningKey as CryptoKey | undefined;
+    /* c8 ignore next 3 */
+    if (clientPrivateKey && !(clientPrivateKey instanceof CryptoKey)) {
+      clientPrivateKey = await jose.importPKCS8<CryptoKey>(clientPrivateKey, clientAssertionSigningAlg || 'RS256');
+    }
+    const response = await oauth.authorizationCodeGrantRequest(
```
Comment (Widcket) [inline src/auth0-session/client/edge-client.ts:137]: Can this call throw an exception? If so, should it be caught, wrapped, and rethrown?

File: src/auth0-session/client/edge-client.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+      redirectUri,
+      checks.code_verifier as string,
+      {
+        additionalParameters: extras.exchangeBody,
+        ...(clientPrivateKey && { clientPrivateKey }),
+        ...this.httpOptions()
+      }
+    );
+
+    const result = await oauth.processAuthorizationCodeOpenIDResponse(
```
Comment (Widcket) [inline src/auth0-session/client/edge-client.ts:150]: Same as above.

File: src/auth0-session/client/edge-client.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+      oidcLogoutUrl.searchParams.set(key, value);
+    });
+
+    oidcLogoutUrl.searchParams.set('client_id', this.config.clientID);
+    return oidcLogoutUrl.toString();
+  }
+
+  async userinfo(accessToken: string): Promise<Record<string, unknown>> {
+    const [as, client] = await this.getClient();
+    const response = await oauth.userInfoRequest(as, client, accessToken, this.httpOptions());
```
Comment (Widcket) [inline src/auth0-session/client/edge-client.ts:204]: Can this call throw an exception? If so, should it be caught, wrapped, and rethrown?

File: src/auth0-session/client/edge-client.ts [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+    return oauth.processUserInfoResponse(as, client, oauth.skipSubjectCheck, response);
+  }
+
+  async refresh(refreshToken: string, extras: { exchangeBody: Record<string, any> }): Promise<TokenEndpointResponse> {
+    const [as, client] = await this.getClient();
+    const res = await oauth.refreshTokenGrantRequest(as, client, refreshToken, {
```
Comment (Widcket) [inline src/auth0-session/client/edge-client.ts:211]: Can this call throw an exception? If so, should it be caught, wrapped, and rethrown?

File: src/auth0-session/client/edge-client.ts [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+      additionalParameters: extras.exchangeBody,
+      ...this.httpOptions()
+    });
+    const result = await oauth.processRefreshTokenResponse(as, client, res);
```
Comment (Widcket) [inline src/auth0-session/client/edge-client.ts:215]: Same as above.

File: src/auth0-session/handlers/login.ts [hunk 9]
```diff
@@ -65,27 +62,31 @@ export default function loginHandlerFactory(
...
-    assert(
-      validResponseTypes.includes(authParams.response_type as string),
-      `response_type should be one of ${validResponseTypes.join(', ')}`
-    );
-    assert(/\bopenid\b/.test(authParams.scope as string), 'scope should contain "openid"');
+    if (!validResponseTypes.includes(authParams.response_type as string)) {
+      throw new Error(`response_type should be one of ${validResponseTypes.join(', ')}`);
+    }
+    if (!/\bopenid\b/.test(authParams.scope as string)) {
+      throw new Error('scope should contain "openid"');
```
Comment (Widcket) [inline src/auth0-session/handlers/login.ts:88]: Should these errors have a more specific type?

File: src/edge.ts [hunk 10]
```diff
@@ -28,39 +29,57 @@ const genId = () => {
     .join('');
 };
 
-function getInstance(params?: ConfigParameters): Auth0Edge {
+const telemetry = { name: 'nextjs-auth0', version };
```
Comment (Widcket) [inline src/edge.ts:32]: Should this be moved to some common helper, as it's also used in https://github.com/auth0/nextjs-auth0/blob/141ec7bff57684747e99baddf02840706adda33b/src/index.ts#L26?

File: src/utils/errors.ts [hunk 11]
```diff
@@ -110,7 +108,7 @@ export class AccessTokenError extends AuthError {
 /**
  * @ignore
  */
-export type HandlerErrorCause = Error | AuthError | HttpError;
+export type HandlerErrorCause = Error | AuthError | (Error & { status: number; statusCode: number });
```
Comment (Widcket) [inline src/utils/errors.ts:111]: Should `Error & { status: number; statusCode: number }` be its own type?

File: example-app/app/nav.tsx (hunk 1 above)
Comment (adamjmcgrath) [reply example-app/app/nav.tsx:45]: I was just moving this over from the old example app, I could change them - but I'd rather do that in another PR

File: src/auth0-session/client/edge-client.ts (hunk 3 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:188]: None of our configuration errors have a type, they just need to tell the developer they've configured the SDK wrong

File: src/auth0-session/client/edge-client.ts (hunk 4 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:137]: This gets caught here https://github.com/auth0/nextjs-auth0/blob/edge-auth-handlers/src/auth0-session/handlers/callback.ts#L61-L79

File: src/auth0-session/client/edge-client.ts (hunk 6 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:204]: This get's caught here https://github.com/auth0/nextjs-auth0/blob/edge-auth-handlers/src/handlers/profile.ts#L237

File: src/auth0-session/client/edge-client.ts (hunk 7 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:211]: This should throw an `AccessTokenError` will fix

File: src/auth0-session/client/edge-client.ts (hunk 8 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:215]: This doesn't throw, if there's an error `oauth.isOAuth2Error(result)` is `true` (which is caught below)

File: src/auth0-session/client/edge-client.ts (hunk 6 above)
Comment (Widcket) [reply src/auth0-session/client/edge-client.ts:204]: In the node handler we're wrapping the errors in our own, more specific error types: https://github.com/auth0/nextjs-auth0/blob/141ec7bff57684747e99baddf02840706adda33b/src/auth0-session/client/node-client.ts#L188

File: src/auth0-session/handlers/login.ts (hunk 9 above)
Comment (adamjmcgrath) [reply src/auth0-session/handlers/login.ts:88]: This is a configuration error too - we don't have a type for these. We could look at changing that, but it would be a lot of changes for not much/any benefit.

File: src/edge.ts (hunk 10 above)
Comment (adamjmcgrath) [reply src/edge.ts:32]: Yeah, I can move this to `shared.ts`

File: src/auth0-session/client/edge-client.ts (hunk 6 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:204]: Not for `userinfo` - but yeah - fair enough, I'll create a `UserInfoError` and add it to the node and edge clients

File: src/utils/errors.ts (hunk 11 above)
Comment (adamjmcgrath) [reply src/utils/errors.ts:111]: Sure - I'll make an `HttpError` type

File: src/auth0-session/client/edge-client.ts (hunk 3 above)
Comment (Widcket) [reply src/auth0-session/client/edge-client.ts:188]: Should this error in particular be a little more descriptive then? If I were a developer not versed in identity and got that error, I think I'd be confused.

File: src/auth0-session/client/edge-client.ts (hunk 4 above)
Comment (Widcket) [reply src/auth0-session/client/edge-client.ts:137]: In the node handler we're wrapping the errors in our own, more specific error types:

https://github.com/auth0/nextjs-auth0/blob/141ec7bff57684747e99baddf02840706adda33b/src/auth0-session/client/node-client.ts#L188

File: tests/auth0-session/handlers/callback.test.ts [hunk 12]
```diff
@@ -436,60 +436,6 @@ describe('callback', () => {
     expect(header.alg).toEqual('RS256');
   });
 
-  it('should use client secret jwt on token endpoint', async () => {
```
Comment (Widcket) [inline tests/auth0-session/handlers/callback.test.ts:439]: Apologies if this is something obvious, why are these tests being removed?

File: tests/fixtures/app-router-helpers.ts [hunk 13]
```diff
@@ -1,23 +1,52 @@
...
+const isEdgeRuntime =
+  // @ts-ignore
+  typeof EdgeRuntime !== 'undefined';
+
+export const initAuth0 = (config: ConfigParameters) => {
+  if (isEdgeRuntime) {
+    return edgeInitAuth0(config);
+  } else {
+    return nodeInitAuth0(config);
+  }
```
Comment (Widcket) [inline tests/fixtures/app-router-helpers.ts:32]:
```suggestion
  }
return nodeInitAuth0(config);
```

File: tests/auth0-session/handlers/callback.test.ts (hunk 12 above)
Comment (adamjmcgrath) [reply tests/auth0-session/handlers/callback.test.ts:439]: Auth0 doesn't support client secret jwt and the edge client doesn't so I've removed support

File: src/auth0-session/client/edge-client.ts [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+        signal: AbortSignal.timeout(this.config.httpTimeout),
+        headers
+      };
+    };
+  }
+
+  private async getClient(): Promise<[oauth.AuthorizationServer, oauth.Client]> {
+    if (this.as) {
+      return [this.as, this.client as oauth.Client];
+    }
```
Comment (Widcket) [inline src/auth0-session/client/edge-client.ts:67]: Not sure how long-lived instances in the Edge runtime can be (apparently [there's still some notion](https://news.ycombinator.com/item?id=31912794) of "cold-start"). If these were short-lived, wouldn't we be calling the discovery endpoint a lot?

Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:67]: We can't do any better than keeping them in memory 🤷 (unless Next.js has some built in distributed cache I haven't heard of)

File: src/auth0-session/client/edge-client.ts (hunk 3 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:188]: Sure - I can improve the message

File: src/auth0-session/client/edge-client.ts (hunk 14 above)
Comment (Widcket) [reply src/auth0-session/client/edge-client.ts:67]: Could we have an option to not use the discovery endpoint?

File: tests/handlers/login-page-router.test.ts [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+import { withoutApi, withApi } from '../fixtures/default-settings';
+import { decodeState } from '../../src/auth0-session/utils/encoding';
+import { setup, teardown } from '../fixtures/setup';
+import { get, getCookie } from '../auth0-session/fixtures/helpers';
+import { Cookie, CookieJar } from 'tough-cookie';
+
+describe('login handler (page router)', () => {
+  afterEach(teardown);
+
+  test('should create a state', async () => {
```
Comment (Widcket) [inline tests/handlers/login-page-router.test.ts:11]:
```suggestion
  test('should create a state, nonce, and code verifier', async () => {
```

File: src/auth0-session/client/edge-client.ts (hunk 14 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:67]: The question of in memory discovery responses not being persisted is not a new thing we've introduced now that it's running on the Edge runtime. Serverless functions that serve the Node runtime also have the same behaviour, so this SDK has had this behaviour since 0.1.

We could ask users to provide enough configuration to skip the Discovery step, sure (we could also do this in express) - but I don't think this is a feature we need to add just because the client can now be served from Edge Serverless functions as well as Node Serverless functions.

Comment (Widcket) [reply src/auth0-session/client/edge-client.ts:67]: Asking because serverless !== edge. As Vercel's architect mentions, "Edge Functions bring another set of trade-offs into the mix".

Comment (Widcket) [reply src/auth0-session/client/edge-client.ts:67]: What was enough/serviceable with serverless might not be with edge functions.

Comment (Widcket) [reply src/auth0-session/client/edge-client.ts:67]: In any case, this probably warrants more investigation and testing with the edge runtime to find out if something more needs to be done.

Comment (Widcket) [reply src/auth0-session/client/edge-client.ts:67]: Since edge functions apparently use Cloudflare Workers, I guess it comes down to the differences between Cloudflare Workers and AWS Lambda.

Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:67]: This SDK is agnostic to a particular edge vendor (Vercel/AWS/Azure/CloudFlare/GCP) - but sure, we can look into how their memory persistence differs between the 2 runtimes. I guess at the very least, Edge functions will be more populous given that their are more Edge locations - so I guess the probability of hitting the same box is less. But on the other hand, the Discovery is served from a CDN, so it's not a huge cost.

File: src/auth0-session/client/edge-client.ts (hunk 4 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:137]: `authorizationCodeGrantRequest` doesn't throw an `IdentityProvider` error (one with an `error` / `error_description` that comes from the IDP) it just returns a fetch response (or throws config errors like, "you forgot the `response_type`")

`processAuthorizationCodeOpenIDResponse` returns an IDP error (not throws) which is thrown on L158

File: src/auth0-session/client/edge-client.ts (hunk 5 above)
Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:150]: See https://github.com/auth0/nextjs-auth0/pull/1269/files/141ec7bff57684747e99baddf02840706adda33b#r1245019748

File: src/auth0-session/client/edge-client.ts (hunk 7 above)
Comment (Widcket) [reply src/auth0-session/client/edge-client.ts:215]: This seems to be missing in https://github.com/auth0/nextjs-auth0/pull/1269/commits/0bb1e4ae418dc5841727bd778ef4af2d47bff586

Comment (adamjmcgrath) [reply src/auth0-session/client/edge-client.ts:215]: Ah yeah, was going to say - this is the same as https://github.com/auth0/nextjs-auth0/pull/1269#discussion_r1245019748 the error from the idp get's thrown from `processRefreshTokenResponse`

**Comment by adamjmcgrath** [comment]
Thanks for the review @Widcket 🙏
